### ISSUE TITLE
Ability to put libraries in a separate directory

### ISSUE THREAD

**Issue by NN---** [dotnet/runtime#51556]
macOS bundle application puts executable in a separate folder and resources in another folder.
.NET dll is considered as a resource since it is not executable which runs.
Unfortunately, .NET published executable expects all files in the same directory.

The following directory structure should be supported:
```
Contents\
 MacOS\
  MyExecutable
Resources\
 libnative.dylib
 System.Runtime.dll
 other dlls.
```

**Comment by NN---** [comment]
@vitek-karas Any update on this ?
Thanks.

**Comment by vitek-karas** [comment]
@NN--- No, this is one of many feature requests, we're trying to prioritize, this one just didn't make it to the top yet.

**Comment by NN---** [comment]
This request is important in macOS due to need of using '--force --deep' flags for codesign which is not recommended.
Also this makes it possible to put executable separately from the libraries in any environment.

**Comment by agocke** [comment]
Would single-file work here? You would have to put the native binaries next to the app, but there would be no other .NET DLLs.

**Comment by NN---** [comment]
Single file does work.

However I needed several assemblies to not be included due to license restriction, therefore I cannot afford single file.

**Comment by agocke** [comment]
Unfortunately, there are significant technical hurdles to moving System.Private.CoreLib.dll away from clr.dll. Therefore, I don't think we'll be able to implement this any time soon.

Out of curiosity, which license prohibits use in single-file? Single-file uses embedding, so it doesn't modify the DLL in any way. It's closer to a unix `tar` file than anything else.

**Comment by NN---** [comment]
For instance LGPL license requires an ability to change the provided file by customer.
If the files are packed inside a single file. it is not possible to replace an LGPL licensed assembly.

**Comment by agocke** [comment]
FWIW https://www.gnu.org/licenses/gpl-faq.html#LGPLStaticVsDynamic seems to imply that you must provide such a layout, but you aren't required to actually ship that version.

**Comment by NN---** [comment]
Ok. I'll check this. Thanks.

Anyway, several .NET APIs work differently in a single file in-memory and extracted environment.
This means I must also make sure both versions work the same.

**Comment by agocke** [comment]
That's true, although the single-file analyzer should now warn about most of those APIs. Based on the last year or so of customer reports, it appears that Assembly.Location is by far the most impactful.

**Comment by NN---** [comment]
Definitely agree about the analyzer, it works well and very helpful, however it analyzes your code and not code of third parties.

Thanks for the answers.

**Comment by NN---** [comment]
Another small difference between single file and extracted files is that if you need a file on the disk when it is extracted you do not have to do anything, while as single file you need to extract it manually from resource for example.

**Comment by agocke** [comment]
Yes, single-file is not particularly suitable if you need direct access to the file. 

And if your license (maybe the LGPL, I don't know) requires being able to build a single-file from binaries only, that is technically possible, but not supported. You could execute the (bundler manually)[https://github.com/dotnet/runtime/blob/6a889d234267a4c96ed21d0e1660dce787d78a38/src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs#L37], and provide exactly the same arguments that were used to construct the bundle on your machine (probably also a specification of what kind of machine you have).
